File: src/lib/aiEngineer/commandDisplay.ts
/** Keep displayed shell clean — titles belong in the card header (`intent`). */

const BANNER_BODY =
  /^\s*[=─\-_*]{2,}[^=─\-_*].*[=─\-_*]{2,}\s*$|^\s*[=─\-_*]{3,}\s*$/;
const EMPTY_ECHO = /^\s*echo\s+(?:-e\s+)?(?:(['"])\s*\1)?\s*$/i;
const ECHO_CMD = /^\s*echo\b/i;
const COMMENT_LINE = /^\s*#/;
const BLANK = /^\s*$/;

export function extractCommandTitle(command: string): string {
  for (const line of command.replace(/\r\n/g, "\n").split("\n")) {
    if (BLANK.test(line)) continue;
    if (COMMENT_LINE.test(line)) return line.replace(/^\s*#\s?/, "").trim();
    break;
  }
  return "";
}

function isDecorativeEchoStatement(stmt: string): boolean {
  const s = stmt.trim();
  if (!s || !ECHO_CMD.test(s)) return false;
  if (EMPTY_ECHO.test(s)) return true;
  const idx = s.toLowerCase().indexOf("echo");
  const rest = s.slice(idx + 4).trim();
  if (!rest) return true;
  const m = rest.match(/^\s*(['"])([\s\S]*?)\1\s*$/);
  if (m) return BANNER_BODY.test(m[2] || "");
  return BANNER_BODY.test(rest);
}

function splitShellStatements(script: string): Array<{ stmt: string; sep: string }> {
  const text = script || "";
  const parts: Array<{ stmt: string; sep: string }> = [];
  let buf = "";
  let i = 0;
  let quote = "";
  while (i < text.length) {
    const ch = text[i]!;
    if (quote) {
      buf += ch;
      if (ch === "\\" && i + 1 < text.length && quote === '"') {
        buf += text[i + 1]!;
        i += 2;
        continue;
      }
      if (ch === quote) quote = "";
      i += 1;
      continue;
    }
    if (ch === "'" || ch === '"') {
      quote = ch;
      buf += ch;
      i += 1;
      continue;
    }
    if (ch === ";" && !(i + 1 < text.length && text[i + 1] === ";")) {
      parts.push({ stmt: buf, sep: ";" });
      buf = "";
      i += 1;
      continue;
    }
    if (text.startsWith("&&", i)) {
      parts.push({ stmt: buf, sep: "&&" });
      buf = "";
      i += 2;
      continue;
    }
    if (text.startsWith("||", i)) {
      parts.push({ stmt: buf, sep: "||" });
      buf = "";
      i += 2;
      continue;
    }
    buf += ch;
    i += 1;
  }
  parts.push({ stmt: buf, sep: "" });
  return parts;
}

function sanitizeLineStatements(line: string): string {
  const chunks = splitShellStatements(line);
  const kept: Array<{ stmt: string; sep: string }> = [];
  for (const { stmt, sep } of chunks) {
    if (isDecorativeEchoStatement(stmt)) continue;
    if (BLANK.test(stmt) && !stmt.trim()) continue;
    kept.push({ stmt, sep });
  }
  if (!kept.length) return "";
  let out = "";
  for (let idx = 0; idx < kept.length; idx++) {
    const { stmt, sep } = kept[idx]!;
    out += idx === 0 ? stmt.replace(/\s+$/, "") : stmt.trim();
    if (idx < kept.length - 1) {
      const joiner = sep || ";";
      out += joiner === ";" ? "; " : ` ${joiner} `;
    }
  }
  return out.trim();
}

export function sanitizeDisplayCommand(command: string): string {
  const lines = command.replace(/\r\n/g, "\n").replace(/\r/g, "\n").split("\n");
  const out: string[] = [];
  let leading = true;
  for (const line of lines) {
    if (leading && BLANK.test(line)) continue;
    if (leading && COMMENT_LINE.test(line)) continue;
    leading = false;
    const cleaned = sanitizeLineStatements(line);
    if (!cleaned) continue;
    if (BLANK.test(cleaned) && (!out.length || BLANK.test(out[out.length - 1]!))) {
      continue;
    }
    out.push(cleaned);
  }
  while (out.length && BLANK.test(out[out.length - 1]!)) out.pop();
  return out.join("\n").trim();
}

File: agent-sidecar/app/harness/command_display.py
"""Keep terminal_exec `command` as executable shell — titles belong in `intent`."""

from __future__ import annotations

import re

# echo whose entire quoted/unquoted payload is a section banner (===…=== / ---…---)
_BANNER_PAYLOAD = re.compile(
    r"""^\s*(?:-e\s+)?(?P<q>['"])(?P<body>.*?)\1\s*$""",
    re.IGNORECASE | re.DOTALL,
)
_BANNER_BODY = re.compile(
    r"""^\s*[=─\-_*]{2,}[^=─\-_*].*[=─\-_*]{2,}\s*$"""
    r"""|^\s*[=─\-_*]{3,}\s*$""",
)
_EMPTY_ECHO = re.compile(
    r"""^\s*echo\s+(?:-e\s+)?(?:(['"])\s*\1)?\s*$""",
    re.IGNORECASE,
)
_COMMENT_LINE = re.compile(r"^\s*#")
_BLANK = re.compile(r"^\s*$")
_ECHO_CMD = re.compile(r"^\s*echo\b", re.IGNORECASE)


def extract_command_title(command: str) -> str:
    """First `# …` comment line, stripped — for UI title when intent is missing."""
    for line in (command or "").splitlines():
        if _BLANK.match(line):
            continue
        if _COMMENT_LINE.match(line):
            return line.lstrip()[1:].strip()
        break
    return ""


def _is_decorative_echo_statement(stmt: str) -> bool:
    s = (stmt or "").strip()
    if not s or not _ECHO_CMD.match(s):
        return False
    if _EMPTY_ECHO.match(s):
        return True
    # echo …args
    rest = s[s.lower().find("echo") + 4 :].strip()
    if not rest:
        return True
    m = _BANNER_PAYLOAD.match(rest)
    if m:
        return bool(_BANNER_BODY.match(m.group("body") or ""))
    # unquoted: echo === foo ===
    if _BANNER_BODY.match(rest):
        return True
    return False


def _split_shell_statements(script: str) -> list[tuple[str, str]]:
    """Split on top-level `;` / `&&` / `||`, respecting quotes and escapes.

    Returns list of (statement, trailing_separator) where trailing_separator is
    '', ';', '&&', or '||' (whitespace preserved on the separator side minimally).
    """
    text = script or ""
    parts: list[tuple[str, str]] = []
    buf: list[str] = []
    i = 0
    n = len(text)
    quote = ""
    while i < n:
        ch = text[i]
        if quote:
            buf.append(ch)
            if ch == "\\" and i + 1 < n and quote == '"':
                buf.append(text[i + 1])
                i += 2
                continue
            if ch == quote:
                quote = ""
            i += 1
            continue
        if ch in ("'", '"'):
            quote = ch
            buf.append(ch)
            i += 1
            continue
        # top-level separators
        if ch == ";" and not (i + 1 < n and text[i + 1] == ";"):
            parts.append(("".join(buf), ";"))
            buf = []
            i += 1
            continue
        if text.startswith("&&", i):
            parts.append(("".join(buf), "&&"))
            buf = []
            i += 2
            continue
        if text.startswith("||", i):
            parts.append(("".join(buf), "||"))
            buf = []
            i += 2
            continue
        buf.append(ch)
        i += 1
    parts.append(("".join(buf), ""))
    return parts


def _sanitize_line_statements(line: str) -> str:
    """Drop decorative echo statements inside one physical line."""
    chunks = _split_shell_statements(line)
    kept: list[tuple[str, str]] = []
    for stmt, sep in chunks:
        if _is_decorative_echo_statement(stmt):
            continue
        if _BLANK.match(stmt) and not stmt.strip("\t "):
            # pure whitespace statement — skip
            continue
        kept.append((stmt, sep))
    if not kept:
        return ""
    # Rebuild with readable separators
    out: list[str] = []
    for idx, (stmt, sep) in enumerate(kept):
        out.append(stmt.rstrip() if idx == 0 else stmt.strip())
        if idx < len(kept) - 1:
            joiner = sep if sep else ";"
            if joiner == ";":
                out.append("; ")
            else:
                out.append(f" {joiner} ")
    return "".join(out).strip()


def first_executable_statement(command: str) -> str:
    """First non-decorative statement after sanitization — for intent fallback."""
    clean = sanitize_terminal_command(command or "")
    if not clean:
        return ""
    flattened = clean.replace("\n", "; ")
    for stmt, _sep in _split_shell_statements(flattened):
        s = re.sub(r"\s+", " ", stmt).strip()
        if s and not _is_decorative_echo_statement(s):
            return s
    return re.sub(r"\s+", " ", clean).strip()


def sanitize_terminal_command(command: str) -> str:
    """Drop title comments and banner `echo ===…===` / `echo ---…---` statements.

    Works for newline-separated scripts and `;` / `&&` / `||` chains.
    Real echo/printf that print useful data are kept.
    """
    lines = (command or "").replace("\r\n", "\n").replace("\r", "\n").split("\n")
    out: list[str] = []
    leading = True
    for line in lines:
        if leading and _BLANK.match(line):
            continue
        if leading and _COMMENT_LINE.match(line):
            continue
        leading = False
        cleaned = _sanitize_line_statements(line)
        if not cleaned:
            continue
        if _BLANK.match(cleaned) and (not out or _BLANK.match(out[-1])):
            continue
        out.append(cleaned)
    while out and _BLANK.match(out[-1]):
        out.pop()
    return "\n".join(out).strip()
